src/cargo/core/compiler/custom_build.rs

```diff
@@ -536,6 +540,44 @@ impl BuildOutput {
     }
 }
 
+fn prepare_metabuild<'a, 'cfg>(
+    cx: &Context<'a, 'cfg>,
+    unit: &Unit<'a>,
+    deps: &[String],
+) -> CargoResult<()> {
+    let mut output = Vec::new();
+    let available_deps = cx.dep_targets(unit);
+    // Filter out optional dependencies, and look up the actual lib name.
+    let meta_deps: Vec<_> = deps
+        .iter()
+        .filter_map(|name| {
+            available_deps
+                .iter()
+                .find(|u| u.pkg.name().as_str() == name.as_str())
+                .map(|dep| dep.target.crate_name())
+        })
+        .collect();
```

Comment:
As a minor nit that I'm not sure needs fixing: this is O(n^2), because it searches every available_deps for every metadep. If some readily available data structure provides an O(1) mapping from name to dep, I'd suggest using that. If one doesn't exist, though, I don't know that it's worth creating one just for this, given that the common case will have relatively few metabuild entries.